[caffe2][nomnigraph] Allow empty storage for the 'Edge' class.

yyetim opened the pull request:
Summary:
This diff:
- Changes the Graph, Edge, etc. classes to allow a void type.
- Refactors Algorithm.h to get rid of unnecessary int placeholder.

Test Plan:
- There are already both types of uses for Edge, and hence both
  template paths are covered. I.e., all unit tests passed => template
  codes covered.
- For scc: Now have two tests for Tarjan's impl.

bwasti commented on caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h:
```diff
@@ -23,28 +23,24 @@
-template <typename T, typename U = T>
-class Edge : public StorageType<U> {
+template <typename T, typename U>
+class EdgeBase {
```
have you tried applying this "base" paradigm to the StorageType class itself?  might clean things up here.  The duplication of of constructors is a bit messy

yyetim replied:
I am not sure (conceptually) how this would work, though. 

To brain storm: 
Say, C++ figured out all the types, and we have actual template instances: We'd have an empty StorageType class:
class EmptyStorageType {};
Now, at the higher layer, we'd have the Edge class. Let's say, for the void case, it inherits from this:
class EdgeEmpty : public EmptyStorageType, public EdgeBase {...}

Now, for constructors for EdgeEmpty:
case 1) There is a three-arg constructor. In this case, the third argument is going to have the void type, and this is not allowed.
case 2) There is no three-arg constructor: In this case, the non-Empty Edge class needs to have a three-arg constructor. Therefore, I still need to have two classes. To avoid duplicating code, I'd still have an EdgeBase, and we come to the same EdgeBase + 2 Edge version, except now we have an extra empty StorageType class that doesn't contain anything.

Do you see anything that I'm missing?

bwasti replied:
I was thinking `Edge` can remain unchanged and inherit from `StorageType<void>` which is a specialized class that doesn't do anything

bwasti replied:
just had a thought that you can use a parameter pack to get it working.  I'm not entirely sure how the compiler chooses a constructor, to be quite honest...

```
diff --git a/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h b/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
index b9b338f37..f513887ca 100644
--- a/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
+++ b/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
@@ -33,7 +33,7 @@ class Graph;
 template <typename T, typename U = T>
 class Node;

-template <typename T, typename U = T>
+template <typename T, typename U = void>
 class Edge : public StorageType<U> {
  public:
   using NodeRef = typename Graph<T, U>::NodeRef;
@@ -43,8 +43,9 @@ class Edge : public StorageType<U> {
     DEBUG_PRINT("Creating instance of Edge: %p\n", this);
   }

-  Edge(NodeRef tail, NodeRef head, U&& data)
-      : StorageType<U>(std::move(data)), Tail(tail), Head(head) {
+  template <typename... Args>
+  Edge(NodeRef tail, NodeRef head, Args... args)
+      : StorageType<U>(std::forward<Args>(args)...), Tail(tail), Head(head) {
     DEBUG_PRINT("Creating instance of Edge: %p\n", this);
   }

diff --git a/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h b/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
index 0f0ba9545..095e30143 100644
--- a/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
+++ b/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
@@ -35,6 +35,12 @@ class StorageType {
   T Data;
 };

+template <>
+class StorageType<void> {
+public:
+  StorageType() {}
+};
+
 /// \brief This class enables a listener pattern.
 /// It is to be used with a "curious recursive pattern"
 /// i.e. Derived : public Notifier<Derived> {}
```

bwasti commented on caffe2/core/nomnigraph/include/nomnigraph/Representations/NeuralNet.h:
```diff
@@ -241,8 +241,8 @@ class GenericOperator : public NeuralNetOperator {
 
-using NNGraph = nom::Graph<std::unique_ptr<nom::repr::Value>, int>;
-using NNSubgraph = nom::Subgraph<std::unique_ptr<nom::repr::Value>, int>;
+using NNGraph = nom::Graph<std::unique_ptr<nom::repr::Value>>;
```
so clean :smile:

yyetim replied:
Thanks 👍

bwasti commented on caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h:
lets throw in a `static_assert` on the number of types passed in

bwasti commented on caffe2/core/nomnigraph/include/nomnigraph/Graph/TarjansImpl.h:
```diff
@@ -2,10 +2,17 @@
 #error "This should only be included by Graph/Algorithms.h"
```
no need for these lines, thanks for cleaning this up

bwasti commented on caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h:
```diff
@@ -2,10 +2,15 @@
 #error "This should only be included by Graph/Algorithms.h"
```
ibid on deleting these

bwasti reviewed (approved):
looks great -- happy to accept this.  make sure all the tests pass before landing

yyetim committed:
Allow empty storage for the 'Edge' class.

This commit:
- Converts edge storage to an optional type.
- Adds a new test in tarjans_test.
- Refactors related bits in other files.

yyetim commented:
Closing this request: All the tests are currently passing except 1. The build error doesn't seem related.
